### PR TITLE
Surface Pressure Gradient Loss: penalize dp/ds mismatch along surface

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current training loss (MAE + DCT freq) optimizes **point-wise** pressure accuracy and **spectral amplitude** fidelity, but never directly penalizes errors in the **spatial gradient of pressure along the surface** (dp/ds). Since:

1. **Pressure gradient drives lift and separation** — the Kutta-Joukowski theorem and boundary layer equations show that dp/ds determines where flow separates and how much lift is generated. Matching Cp values but not dCp/ds can still produce wrong aerodynamic forces.

2. **The aft foil in tandem configurations has the steepest pressure gradients** — wake impingement from the fore foil creates sharp pressure jumps on the aft leading edge. These are exactly where our p_tan errors concentrate.

3. **DCT freq loss catches global spectral content; gradient loss catches local spatial structure** — they are complementary. DCT penalizes wrong *amplitudes* at each frequency; gradient loss penalizes wrong *transitions* between adjacent nodes.

The hypothesis: adding an auxiliary loss on consecutive surface node pressure differences will explicitly force the model to match the **shape** of the pressure distribution, not just its point values and spectral content, improving p_tan on the OOD NACA6416 aft foil.

**Relationship to in-flight experiments:**
- `#2210` (arc-length surface loss reweighting): reweights existing MAE loss by arc-length spacing — not gradient matching
- `#2197` (curvature loss weighting): reweights by geometric curvature — not actual pressure gradient
- `#2184` (DCT freq loss, merged): penalizes spectral mismatch in frequency domain — complementary, not overlapping

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flags (near the `dct_freq_loss` flags, around line 1023)

```python
dp_ds_loss: bool = False      # auxiliary loss on surface pressure gradients (dp/ds)
dp_ds_weight: float = 0.05    # weight for dp/ds gradient loss
```

### Step 2: Extend the raw-coordinate save block (around line 1665) to also trigger on dp_ds_loss

```python
_raw_x_for_dct = x[:, :, 0].clone() if (cfg.dct_freq_loss or cfg.dp_ds_loss) else None
_raw_saf_for_dct = x[:, :, 2:4].norm(dim=-1) if (cfg.dct_freq_loss or cfg.dp_ds_loss) else None
_raw_tandem_for_dct = (x[:, 0, 22].abs() > 0.01) if (cfg.dct_freq_loss or cfg.dp_ds_loss) else None
```

### Step 3: Add the gradient loss block immediately after the DCT loss block (around line 1999)

```python
# Surface pressure gradient (dp/ds) auxiliary loss
if cfg.dp_ds_loss and model.training:
    _dp_loss = torch.tensor(0.0, device=device)
    _n_foils_dp = 0
    for b in range(B):
        surf_idx_b = is_surface[b].nonzero(as_tuple=True)[0]
        if surf_idx_b.numel() < 4:
            continue
        _is_tan_b = _raw_tandem_for_dct[b].item()
        if _is_tan_b:
            saf_vals = _raw_saf_for_dct[b, surf_idx_b]
            foil_groups = [surf_idx_b[saf_vals <= 0.005], surf_idx_b[saf_vals > 0.005]]
        else:
            foil_groups = [surf_idx_b]
        for foil_surf in foil_groups:
            if foil_surf.numel() < 4:
                continue
            # Sort by raw x-coordinate (consistent with DCT ordering)
            x_coords = _raw_x_for_dct[b, foil_surf]
            sort_order = x_coords.argsort()
            sorted_idx = foil_surf[sort_order]
            # Extract pressure channel (channel 2)
            p_pred = pred[b, sorted_idx, 2]  # [M]
            p_gt = y_norm[b, sorted_idx, 2]  # [M]
            # Finite differences: adjacent pressure differences (dp/ds proxy)
            dp_pred = p_pred[1:] - p_pred[:-1]  # [M-1]
            dp_gt = p_gt[1:] - p_gt[:-1]        # [M-1]
            _dp_loss = _dp_loss + (dp_pred - dp_gt).abs().mean()
            _n_foils_dp += 1
    if _n_foils_dp > 0:
        _dp_loss = _dp_loss / _n_foils_dp
        loss = loss + cfg.dp_ds_weight * _dp_loss
```

### Step 4: Training runs

Run 2 seeds. Report W&B run IDs and final surface MAE metrics in a PR comment.

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/dp-ds-w005-s42" \
  --wandb_group "round12/surface-dp-ds" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --dp_ds_loss --dp_ds_weight 0.05 \
  --seed 42

# Seed 73
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/dp-ds-w005-s73" \
  --wandb_group "round12/surface-dp-ds" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --dp_ds_loss --dp_ds_weight 0.05 \
  --seed 73
```

If both seeds beat baseline, optionally test `--dp_ds_weight 0.02` as a 3rd run.

## Baseline

Current best metrics (PR #2184, DCT Frequency-Weighted Loss, 2-seed avg):

| Metric | Baseline | Target |
|--------|----------|--------|
| p_in   | 13.205   | < 13.205 |
| p_oodc | 7.816    | < 7.816  |
| **p_tan** | **28.502** | **< 28.502** |
| p_re   | 6.453    | < 6.453  |

W&B baseline: `6yfv5lio` (seed 42, p_tan=28.432), `etepxvjc` (seed 73, p_tan=28.572)

```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-dct-freq" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — Surface Pressure Gradient Loss (dp/ds)

### Configuration
- `--dp_ds_loss --dp_ds_weight 0.05`
- All other flags match baseline (DCT freq loss, PCGrad 3-way, surface refine, etc.)

### Results

| Metric | Baseline (2-seed avg) | Seed 42 (`cuehr0b0`) | Seed 73 (`nqkaw1cf`) | 2-seed avg | vs Baseline |
|--------|----------------------|----------------------|----------------------|------------|-------------|
| p_in   | 13.205               | 13.007               | 13.426               | 13.22      | -0.1% ≈ tie |
| p_oodc | 7.816                | 7.600                | 8.000                | 7.80       | +0.2% ≈ tie |
| p_tan  | **28.502**           | 28.600               | 29.900               | **29.25**  | **+2.6% worse** |
| p_re   | 6.453                | 6.300                | 6.600                | 6.45       | ≈ tie |

- W&B runs: `cuehr0b0` (seed 42, epoch 148), `nqkaw1cf` (seed 73, epoch 147)
- Epochs reached: 147-148 (comparable to baseline ~150)

### Analysis

**Mixed result — does not beat baseline on the key metric (p_tan).**

- Seed 42 individually looks promising: beats baseline on p_in, p_oodc, and p_re. But p_tan is slightly worse (28.6 vs 28.5).
- Seed 73 is notably worse on p_tan (29.9 vs 28.5) and p_oodc (8.0 vs 7.8), dragging down the average.
- High variance between seeds suggests the dp/ds gradient signal may be noisy — sorting surface nodes by x-coordinate and taking finite differences creates a proxy for dp/ds that doesn't follow the actual arc-length parameterization.

### Suggested Follow-ups

1. **Arc-length-ordered dp/ds**: Instead of sorting by x-coordinate, compute cumulative arc-length from the node coordinates and sort by that. This would give a more physically meaningful finite difference.
2. **Lower weight**: `--dp_ds_weight 0.02` might reduce the noise while still providing a useful auxiliary signal. Not tested because baseline wasn't beaten.
3. **Smooth gradient loss**: Instead of raw finite differences, use a Gaussian-smoothed version to reduce sensitivity to mesh resolution variations.
4. **Aft-foil-only dp/ds**: Apply the gradient loss only to the aft foil (where p_tan errors concentrate) rather than all surface nodes.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2211 because ADVISOR: Closing this experiment — surface pressure gradient loss (dp/ds) did not beat baseline.

**Results summary:**
- 2-seed avg: p_tan=29.25 vs 28.50 baseline (+2.6% regression). p_in, p_oodc, p_re roughly flat.
- Seed 42 individually was close to baseline, but seed 73 showed large regression (p_tan=29.9), indicating high variance.

**Analysis:** The finite-difference dp/ds proxy sorted by x-coordinate doesn't follow the actual arc-length parametrization of the airfoil. At the leading edge where upper and lower surface nodes overlap in x, the sorting conflates nodes from different surfaces, creating spurious large gradients. This is a fundamental issue with x-coordinate-ordered finite differences on closed airfoil geometries. The DCT freq loss (already merged) operates in frequency space and is more robust to this ordering problem.

Your suggestion about arc-length-ordered dp/ds is insightful — fern is currently testing arc-length-based reweighting (#2210) which addresses part of the mesh ordering issue. The aft-foil-only idea is also good but the underlying signal instability needs resolution first.

Good analysis and clean reporting. I'll assign you a new experiment shortly..
